Cython/Utility/StringTools.c

```diff
@@ -665,30 +693,49 @@ static int __Pyx_PyBytes_Tailmatch(PyObject* self, PyObject* substr,
 
 static int __Pyx_PyBytes_SingleTailmatch(PyObject* self, PyObject* arg,
                                          Py_ssize_t start, Py_ssize_t end, int direction) {
-    const char* self_ptr = PyBytes_AS_STRING(self);
-    Py_ssize_t self_len = PyBytes_GET_SIZE(self);
-    const char* sub_ptr;
+    char* self_ptr;
+    Py_ssize_t self_len;
+    char* sub_ptr;
     Py_ssize_t sub_len;
     int retval;
 
+    #if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030B0000
+    PyObject *converted_arg = NULL;
+    #else
     Py_buffer view;
     view.obj = NULL;
+    #endif
 
-    #if !CYTHON_ASSUME_SAFE_SIZE
-    if (unlikely(self_len < 0)) return -1;
+    #if !(CYTHON_ASSUME_SAFE_MACROS && CYTHON_ASSUME_SAFE_SIZE)
+    if (PyBytes_AsStringAndSize(self, &self_ptr, &self_len) == -1) return -1;
+    #else
+    self_ptr = PyBytes_AS_STRING(self);
+    self_len = PyBytes_GET_SIZE(self);
     #endif
 
     if (PyBytes_Check(arg)) {
+        #if !(CYTHON_ASSUME_SAFE_MACROS && CYTHON_ASSUME_SAFE_SIZE)
+        if (PyBytes_AsStringAndSize(arg, &sub_ptr, &sub_len) == -1) return -1;
+        #else
         sub_ptr = PyBytes_AS_STRING(arg);
         sub_len = PyBytes_GET_SIZE(arg);
-        #if !CYTHON_ASSUME_SAFE_SIZE
-        if (unlikely(sub_len < 0)) return -1;
         #endif
     } else {
+        #if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030B0000
+        // Where buffer protocol is unavailable, just convert to bytes
+        // (which is probably inefficient, but does work)
+        converted_arg = PyBytes_FromObject(arg);
+        if (!converted_arg) return -1;
+        if (PyBytes_AsStringAndSize(converted_arg, &sub_ptr, &sub_len) == -1) {
+            Py_DECREF(converted_arg);
+            return -1;
+        }
```

Comment:
Given that this copying is a somewhat heavy operation just to get a buffer, it might be worth adding another special case for bytearray here. Everything else seems sufficiently less common to just do a copy.